zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java

```diff
@@ -0,0 +1,203 @@
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.Reader;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import org.apache.commons.cli.CommandLine;
+import org.apache.commons.cli.CommandLineParser;
+import org.apache.commons.cli.DefaultParser;
+import org.apache.commons.cli.HelpFormatter;
+import org.apache.commons.cli.Options;
+import org.apache.commons.cli.ParseException;
+import org.apache.zeppelin.jupyter.nbformat.Cell;
+import org.apache.zeppelin.jupyter.nbformat.CodeCell;
+import org.apache.zeppelin.jupyter.nbformat.DisplayData;
+import org.apache.zeppelin.jupyter.nbformat.Error;
+import org.apache.zeppelin.jupyter.nbformat.ExecuteResult;
+import org.apache.zeppelin.jupyter.nbformat.MarkdownCell;
+import org.apache.zeppelin.jupyter.nbformat.Nbformat;
+import org.apache.zeppelin.jupyter.nbformat.Output;
+import org.apache.zeppelin.jupyter.nbformat.RawCell;
+import org.apache.zeppelin.jupyter.nbformat.Stream;
+import org.apache.zeppelin.jupyter.zformat.Note;
+import org.apache.zeppelin.jupyter.zformat.Paragraph;
+import org.apache.zeppelin.jupyter.zformat.Result;
+import org.apache.zeppelin.jupyter.zformat.TypeData;
+
+/**
+ *
+ */
+public class JupyterUtil {
+
+  private static final String TEXT_PLAIN = "text/plain";
+  private static final String IMAGE_PNG = "image/png";
+
+  private final RuntimeTypeAdapterFactory<Cell> cellTypeFactory;
+  private final RuntimeTypeAdapterFactory<Output> outputTypeFactory;
+
+  public JupyterUtil() {
+    this.cellTypeFactory = RuntimeTypeAdapterFactory.of(Cell.class, "cell_type")
+        .registerSubtype(MarkdownCell.class, "markdown").registerSubtype(CodeCell.class, "code")
+        .registerSubtype(RawCell.class, "raw");
+    this.outputTypeFactory = RuntimeTypeAdapterFactory.of(Output.class, "output_type")
+        .registerSubtype(ExecuteResult.class, "execute_result")
+        .registerSubtype(DisplayData.class, "display_data")
+        .registerSubtype(Stream.class, "stream").registerSubtype(Error.class, "error");
+  }
+
+  public Nbformat getNbformat(Reader in) {
+    return getNbformat(in, new GsonBuilder());
+  }
+
+  public Nbformat getNbformat(Reader in, GsonBuilder gsonBuilder) {
+    return getGson(gsonBuilder).fromJson(in, Nbformat.class);
+  }
+
+  public Note getNote(Reader in, String codeReplaced, String markdownReplaced) {
+    return getNote(in, new GsonBuilder(), codeReplaced, markdownReplaced);
+  }
+
+  public Note getNote(Reader in, GsonBuilder gsonBuilder, String codeReplaced,
+      String markdownReplaced) {
+    return getNote(getNbformat(in, gsonBuilder), codeReplaced, markdownReplaced);
+  }
+
+  public Note getNote(Nbformat nbformat, String codeReplaced, String markdownReplaced) {
+    Note note = new Note();
+
+    String name = nbformat.getMetadata().getTitle();
+    if (null == name) {
+      name = "Note converted from Jupyter";
```

Comment:
Good point but what if there's no filename provided? The parameter of this method is Reader which it's hard to get the filename from. I suggest it will be used by library.